rust/tools/authorization-logic/raksha_examples/VisionPipelineAuthLogic

```diff
@@ -0,0 +1,274 @@
+    isPrincipal("ProductIdOutput").
+    isPrincipal("LocalDevice").
+}
+
+"ServiceProvider" says {
+    ownsTag("image_detection_model_tag").
+
+    // ServiceProvider also opts-in to "DataEgressPolicy1"
+    "DataEgressPolicy1" canSay principalX may("egress_to_shopping_service", tagX) :-
+        isPrincipal(principalX), isTag(tagX).
+    // ServiceProvider's preference is NOT to egress:
+    "ServiceProvider" hasPreference("disallowEgress").
+
+    // Delegates TPPR the ability to apply just the model tag anywhere
+    "ThirdPartyPrivacyReviewer" canSay hasTag(x, "image_detection_model_tag") :- 
+        isAccessPath(x).
+
+    // Delegate the right to claim that particles are "Local" to TPPR.
+    // (We can think of the principal "LocalDevice" as a group)
+    "ThirdPartyPrivacyReviewer" canSay principalX canActAs "LocalDevice" :-
+        isPrincipal(principalX).
+
+    // Give permission to do pure computation on any local device
+    "LocalDevice" may("do_pure_computation", "image_detection_model_tag").
+
+    // These are rules we should get rid of in the real thing.
+    // The universe relations probably don't need to be decentralized. We could
+    // have rules that populate a "centralized" universe and passes this 
+    // to all the principals. (Or there could be other ways to deal with this).
+    isPrincipal("VideoDataSource").
+    isPrincipal("ImageDetector").
+    isPrincipal("ImageSelector").
+    isPrincipal("ProductIdOutput").
+}
+
+
+"DataEgressPolicy1" says {
+    // Gathers preferences from both stakeholders:
+    "EndUser" canSay "EndUser" hasPreference(prefX) :-
+        isPreference(prefX).
+    "ServiceProvider" canSay "ServiceProvider" hasPreference(prefX) :-
+        isPreference(prefX).
+
+    // Egressing model data and video data is allowed EVEN if
+    // ServiceProvider says "no" as long as EndUser says "yes"
+    // (Both of these facts are used to make the same one decision about 
+    // egressing a handle that depends on both of these tags)
+    "ProductIdOutput" may("egress_to_shopping_service", "product_id_tag") :-
+        "EndUser" hasPreference("allowEgress").
+    "ProductIdOutput" may("egress_to_shopping_service", 
+        "image_detection_model_tag"). // Doesn't care about ServiceProvider pref
+}
+
+// Similar to policy 1, but we need agreement from BOTH parties:
+"DataEgressPolicy2" says {
+    // Gathers preferences from both stakeholders:
+    "EndUser" canSay "EndUser" hasPreference(prefX) :-
+        isPreference(prefX).
+    "ServiceProvider" canSay "ServiceProvider" hasPreference(prefX) :-
+        isPreference(prefX).
+
+    "ProductIdOutput" may("egress_to_shopping_service", "product_id_tag") :-
+        "EndUser" hasPreference("allowEgress").
+    "ProductIdOutput" may("egress_to_shopping_service",
+        "image_detection_model_tag") :-
+        // This policy DOES care about ServiceProvider's preference:
+        "ServiceProvider" hasPreference("allowEgress:).
```

Comment:
is `"allowedEgress:` a typo? Should it be `"allowedEgress"`?

Reply:
Yes, it's a typo. Will fix shortly.